keps/sig-network/2112-netpol-namespace-name-selector/README.md

```diff
@@ -0,0 +1,393 @@
+  - if `namespaceNames` field is retrieved by an older client which **doesn't** yet support it:
+    - the client doesnt crash (and the plugin doesnt crash, either).
+  - if `namespaceNames` is nil:
+    - the policy should behave identically to any policy made before this field was added.
+    - in other words, there is no "deny all" semantic that can be enforced by this namespace being missing.
+  - if `namespaceNames` is empty:
+    - it has IDENTICAL semantics as if it were nil.
+    - in other words, there is no "allow all" namespaces semantic which corresponds to emptyness
+
+Overall, the `NetworkPolicyPeer` modifications should be unobtrusive with regard to how other fields interplay, regardless of its absence or presence, notwithstanding the fact that it potentially broadens the selection of the `namespaceSelector` field.
+
+## Proposal
+
+In NetworkPolicy specification, inside `NetworkPolicyPeer` specify a new `namespaceNames` field.  
+
+1) One simple way to implement this feature is to modify the NetworkPolicyPeer 
+object, like so.  Since this doesn't involve modifying a go type (i.e. were 
+not replacing a labelSelector with a different type, we expect it 
+to be a cleaner implementation:
+
+```
+type NetworkPolicyPeer struct {
+  // new field
+  namespaceNames []string
+
+  // existing fields...
+  PodSelector *metav1.LabelSelector `json:"podSelector,omitempty" protobuf:"bytes,1,opt,name=podSelector"`
+  NamespaceSelector *metav1.LabelSelector `json:"namespaceSelector,omitempty" protobuf:"bytes,2,opt,name=namespaceSelector"`
+  IPBlock *IPBlock `json:"ipBlock,omitempty" protobuf:"bytes,3,rep,name=ipBlock"`
+}
+```
+
+## Design Details
+
+- Add a new selector to the network policy peer data structure which can switch between allowing a `namespaceNames`, supporting a policy that is expressed like this:
+
+- A list of conventional namespace names (i.e. no regular expressions or any other fancy definining syntaxes)
+
+The "namespaceNames" allow directive would look like so:
+```
+kind: NetworkPolicy
+apiVersion: networking.k8s.io/v1
+metadata:
+  name: mysql-allow-app
+spec:
+  podSelector:
+    matchLabels:
+      app: mysql
+  ingress:
+  - from:
+      namespaceNames:
+      -  my-frontend
+      -  my-frontend-2
+```
+
+- As a more sophisticated example: The following would:
+  - allow all traffic from `podName:xyz` living in `my-frontend` or `my-frontend-2`
+  - deny traffic from `podName:xyz` in `my-frontend-3`, since theres no selector for that namespace
+  - allow things in ipblock 100.1.2.0/16
+
+This is because the presence of EITHER a namespaceNames rule or a namespaceSelector, makes the podSelector subject to an **AND** filter operation alongside the pod selector.
+
+```
+kind: NetworkPolicy
+apiVersion: networking.k8s.io/v1
+metadata:
+  name: mysql-allow-app
+spec:
+  podSelector:
+    matchLabels:
+      app: mysql
+  ingress:
+  - from:
+    - namespaceNames:
```

Comment:
This doesn't gel - NamespaceNames and NamespaceSelector should be mutually exclusive, shouldn't they?  They only exception we offer is that you can define NamespaceSelector and PodSelector, so I assume NamespaceNames and PodSelector would be legit.

Reply:
ah missing bullet, fixing

Reply:
fied now